server/sonar-server/src/main/java/org/sonar/server/platform/ws/MigrateDbSystemWsAction.java

```diff
@@ -0,0 +1,168 @@
+import static org.sonar.server.db.migrations.DatabaseMigration.Status.SUCCEEDED;
+
+/**
+ * Implementation of the {@code migrate_db} action for the System WebService.
+ */
+public class MigrateDbSystemWsAction implements SystemWsAction {
+
+  private static final String UNSUPPORTED_DATABASE_MIGRATION_STATUS = "Unsupported DatabaseMigration status";
+
+  private final DatabaseVersion databaseVersion;
+  private final DatabaseMigration databaseMigration;
+  private final Database database;
+
+  public MigrateDbSystemWsAction(DatabaseVersion databaseVersion,
+    Database database,
+    DatabaseMigration databaseMigration) {
+    this.databaseVersion = databaseVersion;
+    this.database = database;
+    this.databaseMigration = databaseMigration;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    controller.createAction("migrate_db")
+      .setDescription("Migrate the database to match the current version of SonarQube." +
+        "<br/>" +
+        "Sending a POST request to this URL will start the DB migration. " +
+        "It is strongly advised to <strong>make a database backup</strong> before invoking this WS." +
+        "<br/>" +
+        "State values are:" +
+        "<ul>" +
+        "<li>NO_MIGRATION: DB is up to date with current version of SonarQube</li>" +
+        "<li>MIGRATION_NEEDED: DB needs to be upgraded in order for the current version of SonarQube to run</li>" +
+        "<li>MIGRATION_RUNNING: DB migration is under go</li>" +
+        "<li>MIGRATION_SUCCEEDED: DB migration has run and has been successful</li>" +
+        "<li>MIGRATION_FAILED: DB migration has run and failed. SonarQube must be restarted in order to retry a " +
+        "DB migration (optionally after DB has been restored from backup)</li>" +
+        "</ul>")
+      .setSince("5.2")
+      .setPost(true)
+      .setHandler(this)
+      .setResponseExample(Resources.getResource(this.getClass(), "example-migrate.json"));
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    Integer currentVersion = databaseVersion.getVersion();
+    if (currentVersion == null) {
+      throw new IllegalStateException("Version can not be retrieved from Database. Database is either blank or corrupted");
+    }
+
+    if (currentVersion >= DatabaseVersion.LAST_VERSION) {
+      writeResponse(response, databaseMigration);
+    } else if (!database.getDialect().supportsMigration()) {
+      writeNotSupportedResponse(response);
+    } else {
+      switch (databaseMigration.status()) {
+        case RUNNING:
+        case FAILED:
+        case SUCCEEDED:
+          writeResponse(response, databaseMigration);
+          break;
+        case NONE:
+          databaseMigration.startIt();
+          writeResponse(response, databaseMigration);
+          break;
+        default:
+          throw new RuntimeException(UNSUPPORTED_DATABASE_MIGRATION_STATUS);
```

Comment:
We usually use IllegalStateException instead of raw RuntimeException. In any case it will raise an issue on Dory because of https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS00112